password_enhancements.module

```diff
@@ -0,0 +1,484 @@
+          'display_label' => TRUE,
+        ],
+        'weight' => 99,
+      ]);
+  }
+
+  return $fields;
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function password_enhancements_user_presave(EntityInterface $entity) {
+  // Update the last password changed date if the password was changed.
+  if ($entity->bundle() == 'user' && (empty($entity->original) || $entity->original->pass->value != $entity->pass->value)) {
+    $entity->password_enhancements_password_changed_date->setValue(Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_user_login().
+ */
+function password_enhancements_user_login(UserInterface $user): void {
+  // Load password the password policy with the highest priority for the given
+  // user.
+  /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+  $password_policy_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_policy');
+  $policy = $password_policy_storage->loadByRoleAndPriority($user->getRoles());
+
+  /** @var \Drupal\Core\Config\Config $config */
+  $config = Drupal::getContainer()->get('config.factory')->get('password_enhancements.settings');
+
+  // Redirect the user to the password change page if the password was expired
+  // or if it was required by the admin.
+  $expire_seconds = !empty($policy) ? $policy->getExpireDays() * 86400 : 0;
+  $request_time = Drupal::time()->getRequestTime();
+  if ((Drupal::request()->get('_route') == 'user.reset.login' && !empty($config->get('require_password_change'))) || $user->password_enhancements_password_change_required->value || ($expire_seconds > 0 && (int) $user->password_enhancements_password_changed_date->value < $request_time - $expire_seconds)) {
+    Drupal::getContainer()->get('session_manager')
+      ->getBag('attributes')
+      ->getBag()
+      ->set('password_enhancements_password_change_required', TRUE);
+  }
+
+  // Set cookie expiry to the same as the password policy's expiry if it is
+  // set, otherwise leave it intact.
+  // @FIXME: Should we somehow notify the user about the expiration?
+  $params = session_get_cookie_params();
+  if ($expire_seconds > 0) {
+    $lifetime = $expire_seconds - ($request_time - (int) $user->password_enhancements_password_changed_date->value);
+    session_set_cookie_params($lifetime > 0 ? $lifetime : 0, $params['path'], $params['domain'], $params['secure'], $params['httponly']);
+    Drupal::getContainer()->get('session_manager')->regenerate();
+  }
+}
+
+/**
+ * Implements hook_legal_allowed_paths_alter().
+ */
+function password_enhancements_legal_allowed_paths_alter(array &$allowed_paths): void {
+  $allowed_paths[] = '/user/password-change';
+}
+
+/**
+ * Implements hook_config_translation_info_alter().
+ */
+function password_enhancements_config_translation_info_alter(&$info) {
+  $info['password_enhancements_constraint']['class'] = 'Drupal\password_enhancements\ConfigTranslation\ConstraintMapper';
```

Comment:
Import the class instead and use `ConstraintMapper::class`. This lets all (proper) IDEs and code analyzer know that the ConstraintMapper class is being used here. From `'Drupal\password_enhancements\ConfigTranslation\ConstraintMapper';` they may not recognize it.